cpp/src/binaryop/compiled/struct_binary_ops.cuh

```diff
@@ -0,0 +1,136 @@
+    tlhs, trhs, compare_orders, {}, stream};
+  auto outd = column_device_view::create(out, stream);
+  auto optional_iter =
+    cudf::detail::make_optional_iterator<bool>(*outd, nullate::DYNAMIC{out.has_nulls()});
+  auto nullate = nullate::DYNAMIC{has_nested_nulls(tlhs) || has_nested_nulls(trhs)};
+
+  auto tabulate_device_operator = [&](auto device_comparator) {
+    thrust::tabulate(
+      rmm::exec_policy(stream),
+      out.begin<bool>(),
+      out.end<bool>(),
+      device_comparison_functor{optional_iter, is_lhs_scalar, is_rhs_scalar, device_comparator});
+  };
+  is_any_v<BinaryOperator, ops::LessEqual, ops::GreaterEqual>
+    ? tabulate_device_operator(table_comparator.less_equivalent(nullate, comparator))
+    : tabulate_device_operator(table_comparator.less(nullate, comparator));
+}
+
+template <typename PhysicalEqualityComparator =
+            cudf::experimental::row::equality::physical_equality_comparator>
+void apply_struct_equality_op(mutable_column_view& out,
+                              column_view const& lhs,
+                              column_view const& rhs,
+                              bool is_lhs_scalar,
+                              bool is_rhs_scalar,
+                              binary_operator op,
+                              PhysicalEqualityComparator comparator = {},
+                              rmm::cuda_stream_view stream          = cudf::default_stream_value)
+{
+  CUDF_EXPECTS(op == binary_operator::EQUAL || op == binary_operator::NOT_EQUAL,
+               "Unsupported operator for these types");
+
+  auto tlhs = table_view{{lhs}};
+  auto trhs = table_view{{rhs}};
+  auto table_comparator =
+    cudf::experimental::row::equality::two_table_comparator{tlhs, trhs, stream};
+  auto device_comparator =
+    table_comparator.equal_to(nullate::DYNAMIC{has_nested_nulls(tlhs) || has_nested_nulls(trhs)},
+                              null_equality::EQUAL,
+                              comparator);
+
+  auto outd = column_device_view::create(out, stream);
+  auto optional_iter =
+    cudf::detail::make_optional_iterator<bool>(*outd, nullate::DYNAMIC{out.has_nulls()});
+  thrust::tabulate(rmm::exec_policy(stream),
+                   out.begin<bool>(),
+                   out.end<bool>(),
+                   [optional_iter,
+                    is_lhs_scalar,
+                    is_rhs_scalar,
+                    preserve_output = (op != binary_operator::NOT_EQUAL),
+                    device_comparator] __device__(size_type i) {
+                     auto lhs = cudf::experimental::row::lhs_index_type{is_lhs_scalar ? 0 : i};
+                     auto rhs = cudf::experimental::row::rhs_index_type{is_rhs_scalar ? 0 : i};
+                     return optional_iter[i].has_value() and
+                            (device_comparator(lhs, rhs) == preserve_output);
+                   });
+}
+}  // namespace cudf::binops::compiled::detail
```

Comment:
There is a convention is that we should leave a blank line at the end of any file.
```suggestion
}  // namespace cudf::binops::compiled::detail

```